perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -1084,200 +1087,214 @@ protected boolean hasCandidateExistingMember(Candidate candidate, RichMember ric
 	 * Adds the group synchronization process in the groupSynchronizerThreads.
 	 *
 	 * @param group
+	 * @throws InternalErrorException when object group is null
 	 */
-	public void forceGroupSynchronization(PerunSession sess, Group group) throws GroupSynchronizationAlreadyRunningException {
-		// First check if the group is not currently in synchronization process
-		if (groupSynchronizerThreads.containsKey(group.getId()) && groupSynchronizerThreads.get(group.getId()).getState() != Thread.State.TERMINATED) {
-			throw new GroupSynchronizationAlreadyRunningException(group);
+	public void forceGroupSynchronization(PerunSession sess, Group group) throws GroupSynchronizationAlreadyRunningException, InternalErrorException {
+		Utils.notNull(group, "group");
+		//Check if the group is not currently in synchronization process
+		if(poolOfGroupsToBeSynchronized.putJobIfAbsent(group, true)) {
+			log.info("Scheduling synchronization for the group {} by force!", group);
 		} else {
-			// Remove from groupSynchronizerThreads if the thread was terminated
-			if (groupSynchronizerThreads.containsKey(group.getId())) {
-				groupSynchronizerThreads.remove(group.getId());
-			}
-			// Start and run the new thread
-			GroupSynchronizerThread thread = new GroupSynchronizerThread(sess, group);
-			thread.start();
-			log.info("Group synchronization thread started for group {}.", group);
-
-			groupSynchronizerThreads.put(group.getId(), thread);
+			throw new GroupSynchronizationAlreadyRunningException(group);
 		}
 	}
 
 	/**
-	 * Synchronize all groups which have enabled synchronization. This method is run by the scheduler every 5 minutes.
+	 * Start and check threads with synchronization of groups. (max threads is defined by constant)
+	 * It also add new groups to the queue.
+	 * This method is run by the scheduler every 5 minutes.
 	 *
 	 * @throws InternalErrorException
 	 */
 	public void synchronizeGroups(PerunSession sess) throws InternalErrorException {
```

Comment:
Better to make whole method synchronized, because it's not designed to run in parallel.